internals/daemon/api_notices.go

```diff
@@ -0,0 +1,169 @@
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"regexp"
+
+	"github.com/canonical/x-go/strutil"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+)
+
+// Ensure custom keys are in the form "domain.com/key" (but somewhat more restrictive).
+var customKeyRegexp = regexp.MustCompile(
+	`^[a-z0-9]+(-[a-z0-9]+)*(\.[a-z0-9]+(-[a-z0-9]+)*)+(/[a-z0-9]+(-[a-z0-9]+)*)+$`)
+
+const (
+	maxNoticeKeyLength = 256
+	maxNoticeDataSize  = 4 * 1024
+)
+
+type addedNotice struct {
+	ID string `json:"id"`
+}
+
+func v1GetNotices(c *Command, r *http.Request, _ *UserState) Response {
+	query := r.URL.Query()
+
+	typeStrs := strutil.MultiCommaSeparatedList(query["types"])
+	types := make([]state.NoticeType, 0, len(typeStrs))
+	for _, typeStr := range typeStrs {
+		noticeType := state.NoticeType(typeStr)
+		if !noticeType.Valid() {
+			// Ignore invalid notice types (so requests from newer clients
+			// with unknown types succeed).
+			continue
+		}
+		types = append(types, noticeType)
+	}
+
+	keys := strutil.MultiCommaSeparatedList(query["keys"])
+
+	after, err := parseOptionalTime(query.Get("after"))
+	if err != nil {
+		return statusBadRequest(`invalid "after" timestamp: %v`, err)
+	}
+
+	filter := &state.NoticeFilter{
+		Types: types,
+		Keys:  keys,
+		After: after,
+	}
+	var notices []*state.Notice
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	timeout, err := parseOptionalDuration(query.Get("timeout"))
+	if err != nil {
+		return statusBadRequest("invalid timeout: %v", err)
+	}
+	if timeout != 0 {
+		// Wait up to timeout for notices matching given filter to occur
+		ctx, cancel := context.WithTimeout(r.Context(), timeout)
+		defer cancel()
+
+		notices, err = st.WaitNotices(ctx, filter)
+		if errors.Is(err, context.Canceled) {
+			return statusBadRequest("request canceled")
+		}
+		// DeadlineExceeded will occur if timeout elapses; in that case return
+		// an empty list of notices, not an error.
+		if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+			return statusInternalError("cannot wait for notices: %s", err)
+		}
+	} else {
+		// No timeout given, fetch currently-available notices
+		notices = st.Notices(filter)
+	}
+
+	if notices == nil {
+		notices = []*state.Notice{} // avoid null result
+	}
+	return SyncResponse(notices)
+}
+
+func v1PostNotices(c *Command, r *http.Request, _ *UserState) Response {
```

Comment:
It looks to me that we can only ever produce one notice with each post. ```v1PostNotices``` seems to suggest otherwise?

Reply:
I don't actually love these names either, but I've gone with convention. It looks like we normally use the plural form on the `/v1/things` endpoints (eg: `v1PostServices`), and the singular on endpoints like `/v1/changes/{id}` which deal with a single change (eg: `v1PostChange`).

Reply:
OK. Good to know. I am all for conformance, thanks.